apps/apps.c

```diff
@@ -2851,3 +2970,164 @@ void make_uppercase(char *string)
     for (i = 0; string[i] != '\0'; i++)
         string[i] = toupper((unsigned char)string[i]);
 }
+
+#define BIO_TYPE_TEMP_FILE      (25|BIO_TYPE_FILTER)
+static BIO_METHOD *methods_tempfile = NULL;
+
+typedef struct {
+    char *filename;
+    char *tmp_filename;
+} TEMPFILE_CTX;
+
+static int tempfile_new(BIO *b);
+static int tempfile_free(BIO *b);
+static int tempfile_read(BIO *b, char *out, int outl);
+static int tempfile_write(BIO *b, const char *in, int inl);
+static long tempfile_ctrl(BIO *b, int cmd, long num, void *ptr);
+static int tempfile_gets(BIO *b, char *buf, int size);
+static int tempfile_puts(BIO *b, const char *str);
+
+static const BIO_METHOD *BIO_f_tempfile(void)
+{
+    if (methods_tempfile == NULL) {
+        methods_tempfile = BIO_meth_new(BIO_TYPE_TEMP_FILE,
+                                        "Temporary file filter");
+        if (methods_tempfile == NULL
+            || !BIO_meth_set_write(methods_tempfile, tempfile_write)
+            || !BIO_meth_set_read(methods_tempfile, tempfile_read)
+            || !BIO_meth_set_puts(methods_tempfile, tempfile_puts)
+            || !BIO_meth_set_gets(methods_tempfile, tempfile_gets)
+            || !BIO_meth_set_ctrl(methods_tempfile, tempfile_ctrl)
+            || !BIO_meth_set_create(methods_tempfile, tempfile_new)
+            || !BIO_meth_set_destroy(methods_tempfile, tempfile_free))
+            return NULL;
+    }
+    return methods_tempfile;
+}
+
+void bio_tempfile_cleanup(void)
+{
+    BIO_meth_free(methods_tempfile);
+}
+
+static int tempfile_new(BIO *b)
+{
+    TEMPFILE_CTX *temp;
+
+    temp = app_malloc(sizeof(*temp), "temp file");
+    temp->filename = NULL;
+    temp->tmp_filename = NULL;
+    BIO_set_data(b, temp);
+    BIO_set_init(b, 1);
+
+    return 1;
+}
+
+static int tempfile_free(BIO *b)
+{
+    TEMPFILE_CTX *temp;
+
+    if (b == NULL)
+        return 0;
+
+    temp = BIO_get_data(b);
+    if (temp->filename != NULL && temp->tmp_filename != NULL) {
+#if !defined(_WIN32) && !defined(__VMS)
+        rename(temp->tmp_filename, temp->filename);
```

Comment:
Why avoid this for `_WIN32` ?
There is a define near the top of `apps.c` and a function later supporting `rename` on Windows.

Reply:
Ahh, didn't notice that...will fix that